OrbitQt/ConnectToStadiaWidget.cpp

```diff
@@ -0,0 +1,410 @@
+                   &ConnectToStadiaWidget::ReloadInstances);
+
+  s_instances_loading_->addTransition(this, &ConnectToStadiaWidget::ErrorOccurred, s_idle);
+  s_instances_loading_->addTransition(this, &ConnectToStadiaWidget::ReceivedInstances, s_idle);
+
+  // STATE s_instance_selected
+  s_instance_selected->addTransition(ui_->refreshButton, &QPushButton::clicked,
+                                     s_instances_loading_);
+  s_instance_selected->addTransition(ui_->connectButton, &QPushButton::clicked,
+                                     s_waiting_for_creds);
+  s_instance_selected->addTransition(ui_->instancesTableView, &QTableView::doubleClicked,
+                                     s_waiting_for_creds);
+  s_instance_selected->addTransition(this, &ConnectToStadiaWidget::Connect, s_waiting_for_creds);
+
+  // STATE s_waiting_for_creds
+  QObject::connect(s_waiting_for_creds, &QState::entered, this,
+                   &ConnectToStadiaWidget::CheckCredentialsAvailable);
+
+  s_waiting_for_creds->addTransition(this, &ConnectToStadiaWidget::ReceivedSshInfo,
+                                     s_waiting_for_creds);
+  s_waiting_for_creds->addTransition(this, &ConnectToStadiaWidget::ReadyToDeploy, s_deploying);
+  s_waiting_for_creds->addTransition(ui_->instancesTableOverlay, &OverlayWidget::Cancelled,
+                                     s_instance_selected);
+  s_waiting_for_creds->addTransition(this, &ConnectToStadiaWidget::ErrorOccurred,
+                                     s_instances_loading_);
+
+  // STATE s_deploying
+  QObject::connect(s_deploying, &QState::entered, this, &ConnectToStadiaWidget::DeployOrbitService);
+
+  s_deploying->addTransition(this, &ConnectToStadiaWidget::ErrorOccurred, s_instance_selected);
+  s_deploying->addTransition(ui_->instancesTableOverlay, &OverlayWidget::Cancelled,
+                             s_instance_selected);
+  s_deploying->addTransition(this, &ConnectToStadiaWidget::Connected, s_connected_);
+
+  // STATE s_connected
+  QObject::connect(s_connected_, &QState::entered, this, [this] {
+    ui_->instancesTableOverlay->SetStatusMessage(
+        QString{"Connected to %1"}.arg(selected_instance_->display_name));
+  });
+  QObject::connect(s_connected_, &QState::exited, this, &ConnectToStadiaWidget::Disconnect);
+
+  s_connected_->addTransition(ui_->instancesTableOverlay, &OverlayWidget::Cancelled,
+                              s_instance_selected);
+  s_connected_->addTransition(this, &ConnectToStadiaWidget::ErrorOccurred, s_instance_selected);
+}
+
+void ConnectToStadiaWidget::ReloadInstances() {
+  CHECK(ggp_client_ != nullptr);
+  instance_model_.SetInstances({});
+
+  ggp_client_->GetInstancesAsync([this](outcome::result<QVector<Instance>> instances) {
+    OnInstancesLoaded(std::move(instances));
+  });
+}
+
+void ConnectToStadiaWidget::CheckCredentialsAvailable() {
+  CHECK(selected_instance_.has_value());
+
+  const std::string& instance_id = selected_instance_->id.toStdString();
```

Comment:
That works due to the reference lifetime extension rule, but shouldn't be invoked here.

`toStdString` returns an r-value - not a reference. You should store that into a temporary variable.

Reply:
done